pkg/metering/writer.go

```diff
@@ -0,0 +1,216 @@
+	"github.com/pingcap/metering_sdk/storage"
+	meteringwriter "github.com/pingcap/metering_sdk/writer/metering"
+
+	"github.com/tikv/pd/pkg/utils/logutil"
+)
+
+const (
+	flushInterval = time.Minute
+	flushTimeout  = flushInterval / 2
+)
+
+// Collector collects events into records with caller-defined fields.
+type Collector interface {
+	Category() string
+	Collect(data any)
+	Flush() []map[string]any
+}
+
+// Writer is used to as a delegate to collect and flush the metering data to the underlying storage.
+type Writer struct {
+	sync.RWMutex
+	id       string
+	ctx      context.Context
+	cancel   context.CancelFunc
+	wg       sync.WaitGroup
+	provider storage.ObjectStorageProvider
+	inner    *meteringwriter.MeteringWriter
+	// collectors holds category -> collector, allowing callers to customize record fields per category.
+	collectors map[string]Collector
+}
+
+// NewWriter creates a new metering writer to collect and report the metering data to the underlying storage.
+func NewWriter(ctx context.Context, meteringConfig *Config, id string) (*Writer, error) {
+	err := meteringConfig.adjust()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create storage provider
+	providerConfig := &storage.ProviderConfig{
+		Type:   meteringConfig.Type,
+		Bucket: meteringConfig.Bucket,
+		Region: meteringConfig.Region,
+		Prefix: meteringConfig.Prefix,
+		AWS: &storage.AWSConfig{
+			AssumeRoleARN: meteringConfig.RoleARN,
+		},
+	}
+	provider, err := storage.NewObjectStorageProvider(providerConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create storage provider for metering")
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	return &Writer{
+		id:       id,
+		ctx:      ctx,
+		cancel:   cancel,
+		provider: provider,
+		inner: meteringwriter.NewMeteringWriter(
+			provider,
+			config.DefaultConfig().WithLogger(zap.L()),
+		),
+		collectors: make(map[string]Collector),
+	}, nil
+}
+
+// Start starts the metering writer background loop
+func (mw *Writer) Start() {
+	if mw == nil {
+		return
+	}
+	mw.wg.Add(1)
+	go mw.meteringLoop()
+}
+
+// Stop stops the metering writer
+func (mw *Writer) Stop() {
+	if mw == nil {
+		return
+	}
+	if mw.cancel != nil {
+		mw.cancel()
+	}
+	mw.wg.Wait()
+	if mw.inner != nil {
+		mw.inner.Close()
+	}
+	log.Info("metering writer stopped")
+}
+
+// RegisterCollector registers a collector for a category. Overwrites existing if any.
+func (mw *Writer) RegisterCollector(collector Collector) {
+	if mw == nil {
+		return
+	}
+	mw.Lock()
+	defer mw.Unlock()
+	if mw.collectors == nil {
+		mw.collectors = make(map[string]Collector)
+	}
+	mw.collectors[collector.Category()] = collector
+}
+
+// meteringLoop runs the background loop to flush metering data periodically.
+func (mw *Writer) meteringLoop() {
+	defer logutil.LogPanic()
+	defer mw.wg.Done()
+
+	now := time.Now()
+	// Truncate to the nearest minute and add the interval to get the next flush time.
+	next := now.Truncate(flushInterval).Add(flushInterval)
```

Comment:
What's the purpose?

Reply:
This is the conclusion after discussion with the metering team: we need to write metering data at the minute level and ensure it carries the correct timestamp, so we need to perform a round-like operation on the time to guarantee accuracy.